tao_compiler/mlir/disc/transforms/disc_convert_fake_quant_op.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "tensorflow/compiler/mlir/disc/IR/hlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+
+#define DEBUG_TYPE "disc-convert-fake-quant-op"
+
+// This file implements the logic to convert fake_quant annotated graph to
+// the real quantized version. An example is shown as following.
+//
+// convert:
+//  const -> fake_quant
+//                      \
+//  input -> fake_quant -> gemm -> fake_quant ->
+// to
+//  const -> quantize
+//                     \
+//  input -> quantize -> quantized_gemm -> dequantize ->
+
+namespace mlir {
+namespace disc_ral {
+namespace {
+
+struct QuantizedDynamicConvOpConverter
+    : public OpRewritePattern<mhlo_disc::FakeQuantOp> {
+  using OpRewritePattern<mhlo_disc::FakeQuantOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo_disc::FakeQuantOp op,
+                                PatternRewriter& rewriter) const override {
+    auto convOp = op.input().getDefiningOp<mhlo::DynamicConvOp>();
+    if (!convOp) return failure();
+
+    auto inputFakeQuantOp =
+        convOp.lhs().getDefiningOp<mhlo_disc::FakeQuantOp>();
+    auto weightFakeQuantOp =
+        convOp.rhs().getDefiningOp<mhlo_disc::FakeQuantOp>();
+    if (!inputFakeQuantOp || !weightFakeQuantOp) return failure();
+    if (inputFakeQuantOp.use_signed() != weightFakeQuantOp.use_signed() ||
+        inputFakeQuantOp.num_bits() != weightFakeQuantOp.num_bits()) {
+      return failure();
```

Comment:
nit: maybe some error message?

Reply:
It's just not supported by this pattern, not an error though.